### PR TITLE
Fix:  Make path dependencies with the same name stays locked

### PR DESCRIPTION
**Pull request by linyihai**
### What does this PR try to resolve?
Fixes: https://github.com/rust-lang/cargo/issues/13405

This is a workround based on https://github.com/rust-lang/cargo/issues/13405#issuecomment-1930496807

### How should we test and review this PR?
first commit will pass, second commit fixed it and update test.

### CODE REVIEW

**Comment by linyihai** [comment]
@ehuss any chance for some feedback?  I'm not sure I'm doing the right thing.

File: src/cargo/core/resolver/encode.rs [hunk 1]
```diff
@@ -404,11 +412,42 @@ impl EncodableResolve {
...
             version,
             HashMap::new(),
-        ))
+        ));
+
+        fn get_source_id<'a>(
+            path_deps: &'a HashMap<String, HashMap<semver::Version, SourceId>>,
+            pkg: &'a EncodableDependency,
+        ) -> Option<&'a SourceId> {
+            {
```
Comment (ehuss) [inline src/cargo/core/resolver/encode.rs:421]: This looks like it has an extraneous block here. Can that be removed?

File: src/cargo/core/resolver/encode.rs [hunk 2]
```diff
@@ -404,11 +412,42 @@ impl EncodableResolve {
...
+                }) {
+                    Some(source_id)
+                } else {
+                    // Next look for cases where the name is the same
+                    path_deps.iter().find_map(|(name, version_source)| {
+                        if name == &pkg.name {
+                            return version_source.values().next();
+                        }
+                        None
+                    })
```
[the remark this answers is not in the document]
Comment (linyihai) [reply src/cargo/core/resolver/encode.rs]: > Can you say more about why this was chosen, and maybe investigate if there is a better way to treat it?|

Please forget my previous writing, I submitted a more reasonable version, please check again sometime.

File: src/cargo/core/resolver/encode.rs [hunk 3]
```diff
@@ -404,11 +412,33 @@ impl EncodableResolve {
...
+        ) -> Option<&'a SourceId> {
+            path_deps.iter().find_map(|(name, version_source)| {
+                if name == &pkg.name {
+                    // If there are multiple candidates for the same name, it needs to be determined by combining versions (See #13405).
+                    if version_source.len() > 1 {
+                        if let Ok(pkg_version) = pkg.version.parse::<semver::Version>() {
+                            if let Some(source_id) = version_source.get(&pkg_version) {
+                                return Some(source_id);
+                            }
+                        }
```
Comment (ehuss) [inline src/cargo/core/resolver/encode.rs:429]: I'm still a little concerned about the behavior here when there are multiple versions, but none of them match. There are two concerns:

* This randomly chooses one of the packages, which is not great for consistency.
* Depending on what changed and which random package it picks, you can end up with two packages with the same SourceId.

I'm still wondering if it is appropriate to return `None` here if there are multiple versions, but none of them match. For example, something like:

```suggestion
                        }
                        
                        return None;
```

Does that make sense?

**Comment by ehuss** [comment]
Here is a test that illustrates the problem with having multiple versions:

```rust
#[cargo_test]
fn same_name_version_changed() {
    let p = project()
        .file(
            "Cargo.toml",
            r#"
                [package]
                name = "foo"
                version = "1.0.0"
                edition = "2021"

                [dependencies]
                foo2 = { path = "foo2", package = "foo" }
            "#,
        )
        .file("src/lib.rs", "")
        .file(
            "foo2/Cargo.toml",
            r#"
                [package]
                name = "foo"
                version = "2.0.0"
                edition = "2021"
            "#,
        )
        .file("foo2/src/lib.rs", "")
        .build();

    p.cargo("tree")
        .with_stderr("[LOCKING] 2 packages to latest compatible versions")
        .with_stdout(
            "\
foo v1.0.0 ([ROOT]/foo)
└── foo v2.0.0 ([ROOT]/foo/foo2)
",
        )
        .run();

    p.change_file(
        "foo2/Cargo.toml",
        r#"
            [package]
            name = "foo"
            version = "2.0.1"
            edition = "2021"
        "#,
    );
    p.cargo("tree")
        .with_stderr(
            "\
[LOCKING] 1 package to latest compatible version
[ADDING] foo v2.0.1 ([ROOT]/foo/foo2)
",
        )
        .with_stdout(
            "\
foo v1.0.0 ([ROOT]/foo)
└── foo v2.0.1 ([ROOT]/foo/foo2)
",
        )
        .run();
}
```

With this PR as-is, this test will randomly fail.

I'd recommend adding this test to the `path.rs` testsuite module.

**Comment by linyihai** [comment]
Oh on, the ci faild due to 
```
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    bench::many_similar_names

test result: FAILED. 3326 passed; 1 failed; 7 ignored; 0 measured; 0 filtered out; finished in 390.01s

error: test failed, to rerun pass `-p cargo --test testsuite`
```

I guess [this pr ](https://github.com/rust-lang/cargo/pull/13901) merge can solve this

**Comment by ehuss** [comment]
Thanks! I moved the suggested test to `path.rs` since it is unrelated to patching.  I'm still not 100% confident with that final `None` return, but I can't think of any real alternatives.

@bors r+

**Pull request merged**